src/runtime/uv/tcp.cpp

```diff
@@ -0,0 +1,532 @@
+    }
+
+    event_loop_unlock(&global_ev);
+
+    lean_inc(promise);
+    return lean_io_result_mk_ok(promise);
+}
+
+extern "C" LEAN_EXPORT lean_obj_res lean_uv_tcp_recv(b_obj_arg socket) {
+    lean_uv_tcp_socket_object * tcp_socket = lean_to_uv_tcp_socket(socket);
+
+    if (tcp_socket->m_byte_array != NULL) {
+        std::string err = std::string("failed to read from tcp_socket: parallel read is not allowed!");
+        return io_result_mk_error(err.c_str());
+    }
+
+    lean_object * promise = create_promise();
+    tcp_socket->m_promise_read = promise;
+
+    event_loop_lock(&global_ev);
+
+    // The event loop owns the socket.
+    lean_inc(socket);
+
+    uv_read_start((uv_stream_t*)tcp_socket->m_uv_tcp, [](uv_handle_t *handle, size_t suggested_size, uv_buf_t *buf) {
+        lean_uv_tcp_socket_object * tcp_socket = lean_to_uv_tcp_socket((lean_object*)handle->data);
+
+        tcp_socket->m_byte_array = lean_alloc_sarray(1, 0, suggested_size);
+
+        buf->base = (char*)lean_sarray_cptr(tcp_socket->m_byte_array);
+        buf->len = suggested_size;
+
+    }, [](uv_stream_t *stream, ssize_t nread, const uv_buf_t *buf) {
+        lean_uv_tcp_socket_object * tcp_socket = lean_to_uv_tcp_socket((lean_object*)stream->data);
+        lean_object * promise = tcp_socket->m_promise_read;
+
+        if (nread >= 0) {
+            lean_sarray_set_size(tcp_socket->m_byte_array, nread);
+            resolve_promise(promise, mk_ok_except(tcp_socket->m_byte_array));
+        } else if (nread < 0) {
+            resolve_promise(promise, mk_err_except(lean_mk_io_user_error(mk_string(uv_strerror(nread)))));
+            lean_dec(tcp_socket->m_byte_array);
+        }
+
+        tcp_socket->m_promise_read = NULL;
+        tcp_socket->m_byte_array = NULL;
+
+        uv_read_stop(stream);
+
+        // The event loop does not owns the object anymore.
+        lean_dec((lean_object*)stream->data);
+    });
+
+    event_loop_unlock(&global_ev);
+
+    lean_inc(promise);
+    return lean_io_result_mk_ok(promise);
+}
+
+extern "C" LEAN_EXPORT lean_obj_res lean_uv_tcp_bind(b_obj_arg socket, b_obj_arg addr) {
+    lean_uv_tcp_socket_object * tcp_socket = lean_to_uv_tcp_socket(socket);
+
+    sockaddr addr_ptr;
+    lean_socket_addr_to_sockaddr(addr, &addr_ptr);
+
+    event_loop_lock(&global_ev);
+
+    int result = uv_tcp_bind(tcp_socket->m_uv_tcp, (const struct sockaddr *)&addr_ptr, 0);
+
+    event_loop_unlock(&global_ev);
+
+    if (result < 0) {
+        std::string err = std::string("failed to bind tcp_socket: ") + uv_strerror(result);
+        return io_result_mk_error(err.c_str());
+    }
+
+    return lean_io_result_mk_ok(lean_box(0));
+}
+
+extern "C" LEAN_EXPORT lean_obj_res lean_uv_tcp_listen(b_obj_arg socket, int32_t backlog) {
+    lean_uv_tcp_socket_object * tcp_socket = lean_to_uv_tcp_socket(socket);
+
+    event_loop_lock(&global_ev);
+
+    // The event loop owns the socket.
+    lean_inc(socket);
```

Comment:
While it is probably right that we need to `lean_inc` socket here this feels a bit dangerous. After all there is no corresponding `lean_dec` so `Tcp.listen` is effectively going to act like a repeating `Timer` with the slight difference that we do not really have a function to stop listening. This is definitely a problem we nee to have a solution for before merging.